docs/design/018-new-http-api.md

```diff
@@ -0,0 +1,486 @@
+
+
+#### Sockets
+
+A Socket represents the file or network socket over which client/server or peer communication happens.
+
+It can be of three types:
+- `tcp`: a stream-oriented network socket using the Transmission Control Protocol.
+- `udp`: a message-oriented network socket using the User Datagram Protocol.
+- `ipc`: a mechanism for communicating between processes on the same machine, typically using files.
+
+The Socket state can either be _active_—meaning connected for a TCP socket, bound for a UDP socket, or open for an IPC socket—, or _inactive_—meaning disconnected, unbound, or closed, respectively.
+
+##### Example
+
+- TCP:
+```javascript
+import { TCP } from 'k6/x/net';
+
+export default async function () {
+  const socket = await TCP.open('192.168.1.1:80', {
+                            // default      | possible values
+    ipVersion: 0,           // 0            | 4 (IPv4), 6 (IPv6), 0 (both)
+    keepAlive: true,        // false        |
+    lookup: null,           // dns.lookup() |
+    proxy: 'myproxy:3030',  // ''           |
+  });
+  console.log(socket.active); // true
+
+  // Writing directly to the socket.
+  // Requires TextEncoder implementation, otherwise typed arrays can be used as well.
+  await socket.write(new TextEncoder().encode('GET / HTTP/1.1\r\n\r\n'));
+
+  // And reading...
+  socket.on('data', (data) => {
+    console.log(`received ${data}`);
+    socket.close();
+  });
+
+  await socket.done();
+}
+```
+
+- UDP:
+```javascript
+import { UDP } from 'k6/x/net';
+
+export default async function () {
+  const socket = new UDP.open('192.168.1.1:9090');
+
+  await socket.write(new TextEncoder().encode('GET / HTTP/1.1\r\n\r\n'));
+}
+```
+
+- IPC:
+```javascript
+import { IPC } from 'k6/x/net';
+import { Client } from 'k6/x/net/http';
+
+export default async function () {
+  const socket = await IPC.open('/tmp/unix.sock');
+
+  console.log(socket.file.path); // /tmp/unix.sock
+
+  // The HTTP client supports communicating over a Unix socket.
+  const client = new Client({
+    socket: socket,
+  });
+  await client.get('http://unix/get');
+}
+```
+
+#### Client
+
+An HTTP Client is used to communicate with an HTTP server.
+
+##### Examples
+
+- Using a client with default transport settings, and making a GET request:
+```javascript
+import { Client } from 'k6/x/net/http';
+
+export default async function () {
+  const client = new Client();
+  const response = await client.get('https://httpbin.test.k6.io/get');
+  const jsonData = await response.json();
```

Comment:
```suggestion
  const jsonData = await response.json();
```

This part of the current HTTP API is probably not a good design to copy. Why should a generic HTTP response have a `.json()` method? 

It's fine to have something that wraps a regular `Client` and works only with JSON, but a generic `http.Client` should probably not need to know what JSON is.

Reply:
`http.Client` wouldn't know what JSON is, but I don't see why `http.Response` couldn't have helper methods to parse the response body and return it as various objects.

The [web Fetch API has a `Response` with such methods](https://developer.mozilla.org/en-US/docs/Web/API/Response#instance_methods), and [so does Deno's implementation](https://deno.com/deploy/docs/runtime-response#methods). Regardless if we end up implementing Fetch or not, this is very convenient, and it would be a UX regression if we decide to remove them.

Reply:
Fetch and Deno might have this, but [Go](https://pkg.go.dev/net/http#Response), [axios](https://github.com/axios/axios#response-schema) and [got](https://github.com/sindresorhus/got/blob/main/documentation/3-streams.md#response-2) don't. We should consider what makes sense for us. 

These are my thoughts on the topic, in roughly a descending order of certainty:
1. We definitely shouldn't have a [`.json([selector])` method like the current one](https://k6.io/docs/javascript-api/k6-http/response/response-json/), where we have a `selector` as the argument. That should be a separate JSONPath API that is not tied directly together to the core HTTP API, it should probably work with any string or buffer.
2.  It's fine to have a `Fetch` polyfill/wrapper that has a `.json()` method, built on top of our generic HTTP `Client` (either in Go or in JS!). This solves some of the UX problem of a generic API not having one.
3. It's probably a good idea to have a dedicated custom `JSONClient` type built on top of the generic `Client`, which will be well suited for easily working with JSON REST APIs. Including potentially automatically marshaling request bodies that are objects and maybe even automatically universalizing responses. Maybe even with some JSONPath integration :thinking: This provides all of the UX benefits, but without anything magical and with a clear separation of concerns (i.e. the composable approach).
4. The default generic HTTP `Client` should not make any guesses about the content of the HTTP requests and responses it is handling. This gives us more room for optimization, a clear separation of concerns, and a more consistent UX.
5. We probably shouldn't have a `.json()` method (without any arguments) on the generic HTTP `Client`. If someone _really_ wants to use the generic k6 http `Client` for one-off request with a JSON response, then `JSON.parse(resp.body)` is not that much worse than `resp.json()`...
6. If we decide to add `.json() `, are we also going to add to port the current [`.html()`](https://k6.io/docs/javascript-api/k6-http/response/response-html/)? If not, then why not? And are also going to have `.xml()` or `.protobuf()`?

Reply:
Fetch and Deno might have this, but [Go](https://pkg.go.dev/net/http#Response), [axios](https://github.com/axios/axios#response-schema) and [got](https://github.com/sindresorhus/got/blob/main/documentation/3-streams.md#response-2) don't. We should consider what makes sense for us. 

These are my thoughts on the topic, in roughly a descending order of certainty:
1. We definitely shouldn't have a [`.json([selector])` method like the current one](https://k6.io/docs/javascript-api/k6-http/response/response-json/), where we have a `selector` as the argument. That should be a separate JSONPath API that is not tied directly together to the core HTTP API, it should probably work with any string or buffer.
2.  It's fine to have a `Fetch` polyfill/wrapper that has a `.json()` method, built on top of our generic HTTP `Client` (either in Go or in JS!). This solves some of the UX problem of a generic API not having one.
3. It's probably a good idea to have a dedicated custom `JSONClient` type built on top of the generic `Client`, which will be well suited for easily working with JSON REST APIs. Including potentially automatically marshaling request bodies that are objects and maybe even automatically universalizing responses. Maybe even with some JSONPath integration :thinking: This provides all of the UX benefits, but without anything magical and with a clear separation of concerns (i.e. the composable approach).
4. The default generic HTTP `Client` should not make any guesses about the content of the HTTP requests and responses it is handling. This gives us more room for optimization, a clear separation of concerns, and a more consistent UX.
5. We probably shouldn't have a `.json()` method (without any arguments) on the generic HTTP `Client`. If someone _really_ wants to use the generic k6 http `Client` for one-off request with a JSON response, then `JSON.parse(resp.body)` is not that much worse than `resp.json()`...
6. If we decide to add `.json() `, are we also going to add to port the current [`.html()`](https://k6.io/docs/javascript-api/k6-http/response/response-html/)? If not, then why not? And are also going to have `.xml()` or `.protobuf()`?

Reply:
> Fetch and Deno might have this, but Go, axios and got don't.

I wasn't familiar with got, looks nice :thumbsup: It doesn't implement `json()` directly on `Response`, but [does on the `Promise`](https://github.com/sindresorhus/got/blob/main/documentation/1-promise.md#goturl-string--url-options-optionsinit-defaults-options). It even supports [the `json` property for `POST` requests](https://github.com/sindresorhus/got#json-mode).

BTW, there's another HTTP lib from the same team called [ky](https://github.com/sindresorhus/ky). :sweat_smile: This one apparently targets Deno, but it follows similar design principles as got.

[Axios actually parses JSON bodies by default](https://masteringjs.io/tutorials/axios/json) :smile: But I don't think we should look at it for inspiration, considering it's quite old at this point, and has been superseeded by other libs. It's still very popular and featureful, but also has [many limitations](https://github.com/sindresorhus/got#comparison).

Go shouldn't be an inspiration for the JS API. As stated in the design goals, we want to make things familiar to JS devs, not Go devs. We can borrow certain concepts to make the API composable and whatnot, but refering to it directly would only pollute our way of thinking about idiomatic JS.

> We should consider what makes sense for us.

If by "us" you mean "k6 users", then agreed. These convenience methods are all part of offering a good experience for JS developers. I don't see why that's so controversial.


To address some of your points:

- Sure, the Selector API makes sense as a separate JSONPath API. `json()` should just return the body as an object, as it does in the other JS libs.

- Having purpose-built `Client` implementations just for sending different headers, interpreting requests and responses, seems like overkill, and might be too limiting in some cases. What if you want to send JSON but receive binary data, or any such weird combination we didn't predict? Having a single `Client` that offers some helper mechanisms for the most common use cases adds a fairly low overhead, which can be ignored for those who don't need it. I.e. everyone can just ignore the `json` property, and use `body` instead.

- I'm not sure why you keep mentioning this, but `Client` itself wouldn't have a `json()` method. It would return a `Promise<Response>` which would have it.

- I don't see why we couldn't have `html()` either. The point of these methods are _convenience_ for the most common use cases. If there are good reasons to include `xml()` or `protobuf()`, then we should consider them as well.

  Are you suggesting that we have an `HTMLClient`, a `ProtobufClient` or an `XMLClient` instead?

  Separate client implementations make sense for truly different protocols. So it probably makes sense to have a `GRPCClient` that extends the base `Client`, or maybe a `SOAPClient` :smile:, but not just to avoid these simple payload handling helpers.

Reply:
> Go shouldn't be an inspiration for the JS API. As stated in the design goals, we want to make things familiar to JS devs, not Go devs. We can borrow certain concepts to make the API composable and whatnot, but refering to it directly would only pollute our way of thinking about idiomatic JS.

I mostly agree, but I will push back slightly on this. Go **can** and probably should be an inspiration, as long as the end result is also an idiomatic JavaScript API. If Go has an elegant solution to a problem, and if that solution doesn'tt look strange for JS APIs, is easy to understand, and cleanly solves a bunch of problems, why not adopt it? :confused: Sure, we don't need to copy the Go API directly, but we don't have to avoid its good ideas just be cause they are not JS. Good API design is _somewhat_ universal after all...

> Having purpose-built Client implementations just for sending different headers, interpreting requests and responses, seems like overkill, and might be too limiting in some cases. 

I didn't meed completely separate `Client` implementations. Just wrappers around a generic `Client` with maybe a few extra methods and some pre-defined Event handlers for the same events that you mention in the design doc.

> What if you want to send JSON but receive binary data, or any such weird combination we didn't predict? 

Use the generic `Client`, if it's a one-off, or build your own wrapper (because you should be able to easily do that, since everything is composable).

>  Having a single Client that offers some helper mechanisms for the most common use cases adds a fairly low overhead, which can be ignored for those who don't need it. I.e. everyone can just ignore the json property, and use body instead.

Sure, but having more than one way to do the same thing, especially when it's a simple thing, is generally somewhat of an API design smell. I don't object too strongly to the simple `.json()` method, it's only at point 5 in my list, but I don't see the need. It's also one of these extra things that we can easily add at any future point, but once we add it we can never remove, so I don't see the need to start with it.

> I'm not sure why you keep mentioning this, but Client itself wouldn't have a json() method. It would return a Promise<Response> which would have it.

Sorry, I meant the Response object returned by the generic Client (or, I guess, the Promise that resolves to it).

> I don't see why we couldn't have html() either. The point of these methods are convenience for the most common use cases. If there are good reasons to include xml() or protobuf(), then we should consider them as well.

See the problems that `.html()` causes for the current `k6/http` API... Tightly coupling these is a mistake. 

> Are you suggesting that we have an HTMLClient, a ProtobufClient or an XMLClient instead?

No, that was the point. At best, we may have some helpers and wrappers, but we probably shouldn't. We should make it super easy and efficient for users to build whatever helpers and wrappers. We should provide the flexible and generic toolbox, not try to build a dedicated tool to solve every possible weird use case that exists out there.

Reply:
Hmm actually, I am slightly walking back some of my claims... :sweat_smile:  Thinking about idiomatic `Promise`-based JS APIs, a `.json()` helper on a `Promise` that also returns another `Promise` actually makes some sense :sweat_smile: I still don't like it, but it _probably_ provides enough value to pass muster :thinking: